# 1. Reading CSV files with NumPy

* We learned that NumPy makes it quick and easy to select data, and includes a number of functions and methods that make it easy to calculate statistics across the different axes (or dimensions).

* However, what if we also wanted to find out how many trips were taken in each month? Or which airport is the busiest? For this, we will learn a new technique: Boolean Indexing.

* Use the `numpy.genfromtxt()` function to read files into NumPy ndarrays from csv

* `np.genfromtxt(filename, delimiter=None)`
* filename: A positional argument, usually a string representing the path to the text file to be read.
* delimiter: A named argument, specifying the string used to separate each value.

In [1]:
import numpy as np 
taxi=np.genfromtxt('nyc_taxis.csv',delimiter=',')

In [2]:
taxi_shape=taxi.shape
print(taxi_shape)

(89561, 15)


# 2. Reading CSV files with NumPy Continued

*  when numpy.genfromtxt() reads in a file, it attempts to determine the data type of the file by looking at the values.

In [3]:
taxi.dtype

dtype('float64')

In [4]:
print(taxi[:4])

[[      nan       nan       nan       nan       nan       nan       nan
        nan       nan       nan       nan       nan       nan       nan
        nan]
 [2.016e+03 1.000e+00 1.000e+00 5.000e+00 0.000e+00 2.000e+00 4.000e+00
  2.100e+01 2.037e+03 5.200e+01 8.000e-01 5.540e+00 1.165e+01 6.999e+01
  1.000e+00]
 [2.016e+03 1.000e+00 1.000e+00 5.000e+00 0.000e+00 2.000e+00 1.000e+00
  1.629e+01 1.520e+03 4.500e+01 1.300e+00 0.000e+00 8.000e+00 5.430e+01
  1.000e+00]
 [2.016e+03 1.000e+00 1.000e+00 5.000e+00 0.000e+00 2.000e+00 6.000e+00
  1.270e+01 1.462e+03 3.650e+01 1.300e+00 0.000e+00 0.000e+00 3.780e+01
  2.000e+00]]


NaN is an acronym for Not a Number - it literally means that the value cannot be stored as a number. It is similar to (and often referred to as a) null value, like Python's None constant.

* NaN is most commonly seen when a value is missing, but in this case, we have NaN values because the first line from our CSV file contains the names of each column. NumPy is unable to convert string values like pickup_year into the float64 data type.

* For now, we need to remove this header row from our ndarray. 
* Alternatively, we can pass an additional parameter, skip_header, to the numpy.genfromtxt() function. The skip_header parameter accepts an integer, the number of rows from the start of the file to skip.
* Note that because this integer should be the number of rows and not the index, skipping the first row would require a value of 1, not 0.

In [5]:
taxi=np.genfromtxt('nyc_taxis.csv',delimiter=',',skip_header=1)

In [6]:
taxi.shape

(89560, 15)

# 3. Boolean Arrays

* A boolean array, as the name suggests, is an array of boolean values. Boolean arrays are sometimes called boolean vectors or boolean masks.

In [7]:
a = np.array([1, 2, 3, 4, 5])
a_bool=a<3
a_bool

array([ True,  True, False, False, False])

In [8]:
print(a[a_bool])

[1 2]


# 4. Boolean Indexing with 1D ndarrays

**The boolean array acts as a filter, so that the values corresponding to True become part of the result and the values corresponding to False are removed.**

**(1)use boolean indexing to confirm the number of taxi rides in our data set from the month of January.**

In [11]:
pickup_month=taxi[:,1]
print(pickup_month[:10])

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [16]:
january_bool=pickup_month==1

In [17]:
january_pickups=pickup_month[january_bool]
print(january_pickups[:15])

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


**no of rides in january**

In [18]:
january_pickups.shape

(13481,)

**13481 rides in january**

**(2)use boolean indexing to confirm the number of taxi rides in our data set from the month of December.**

In [23]:
december_bool=pickup_month==float(12)
december_rides=pickup_month[december_bool]
print(december_rides[:4])

[]


In [24]:
december_rides.shape

(0,)

**(3)use boolean indexing to confirm the number of taxi rides in our data set from the month of June.**

In [25]:
june_bool=pickup_month==6
june_rides=pickup_month[june_bool]
print(june_rides[:5])

[6. 6. 6. 6. 6.]


In [26]:
june_rides.shape

(15739,)

**15739 rides in June**

# 5. Boolean Indexing with 2D ndarrays

* When working with 2D ndarrays, you can use boolean indexing in combination with any of the indexing methods.`The only limitation is that the boolean array must have the same length as the dimension you're indexing.`

In [29]:
# calculate the average speed
trip_mph = taxi[:,7] / (taxi[:,8] / 3600)
print(trip_mph.max())

82800.0


In [28]:
# create a boolean array for trips with average
# speeds greater than 20,000 mph
trip_mph_bool = trip_mph > 20000

# use the boolean array to select the rows for
# those trips, and the pickup_location_code,
# dropoff_location_code, trip_distance, and
# trip_length columns
trips_over_20000_mph = taxi[trip_mph_bool,5:9]

print(trips_over_20000_mph)

[[ 2.   2.  23.   1. ]
 [ 2.   2.  19.6  1. ]
 [ 2.   2.  16.7  2. ]
 [ 3.   3.  17.8  2. ]
 [ 2.   2.  17.2  2. ]
 [ 3.   3.  16.9  3. ]
 [ 2.   2.  27.1  4. ]]


* We can see from the last column that most of these are very short rides - all have trip_length values of 4 or less seconds, which does not reconcile with the trip distances, all of which are more than 16 miles.

**(2)examine the rows that have the highest values for the tip_amount column.**

In [30]:
tip_amount = taxi[:,12]
tip_bool=tip_amount>50
top_tips=taxi[tip_bool,5:14]

In [32]:
top_tips[:3]

array([[4.0000e+00, 2.0000e+00, 2.1450e+01, 2.0040e+03, 5.2000e+01,
        8.0000e-01, 0.0000e+00, 5.2800e+01, 1.0560e+02],
       [3.0000e+00, 4.0000e+00, 9.2000e+00, 1.0410e+03, 2.7000e+01,
        1.3000e+00, 5.5400e+00, 6.0000e+01, 9.3840e+01],
       [2.0000e+00, 0.0000e+00, 1.9800e+01, 1.6710e+03, 5.2500e+01,
        1.3000e+00, 5.5400e+00, 5.9340e+01, 1.1868e+02]])

# 6. Assigning Values in ndarrays

`ndarray[location_of_values] = new_value`

### TO DO:
* (1) The value at column index 5 (pickup_location) of row index 28214 is incorrect. Use assignment to change this value to 1          in the taxi_modified ndarray.
* (2) The first column (index 0) contains year values as four digit numbers in the format YYYY (2016, since all trips in our          data set are from 2016). Use assignment to change these values to the YY format (16) in the taxi_modified ndarray.
* (3) The values at column index 7 (trip_distance) of rows index 1800 and 1801 are incorrect. Use assignment to change these          values in the taxi_modified ndarray to the mean value for that column.

In [33]:
taxi_modified=taxi.copy()

In [34]:
# updating specific item

taxi_modified[28214,5]=1

In [35]:
# updating whole column

taxi_modified[:,0]=16

In [36]:
# updating slice of data

taxi_modified[1800:1802,7]=taxi_modified[:,7].mean()

# 7. Assignment Using Boolean Arrays

* Boolean arrays become very powerful when we use them for assignment.

**The boolean array controls the values that the assignment applies to, and the other values remain unchanged.**

### TO DO:
* select the fourteenth column (index 13) in taxi_copy. Assign it to a variable named total_amount.
* For rows where the value of total_amount is less than 0, use assignment to change the value to 0

In [39]:
taxi_copy=taxi.copy()

total_amount=taxi_copy[:,13]
taxi_copy[total_amount<0]=0


## short way
taxi_copy[taxi_copy[:,13]<0]=0

# 8. Assignment Using Boolean Arrays Continued

`bool = array[:, column_for_comparison] == value_for_comparison
array[bool, column_for_assignment] = new_value`

#### In one line
`array[array[:, column_for_comparison] == value_for_comparison, column_for_assignment] = new_value`

## TO DO:
We have created a new copy of our taxi dataset, taxi_modified with an additional column containing the value 0 for every row.

* In our new column at index 15, assign the value 1 if the pickup_location_code (column index 5) corresponds to an airport location, leaving the value as 0 otherwise by performing these three operations:
* For rows where the value for the column index 5 is equal to 2 (JFK Airport), assign the value 1 to column index 15.
* For rows where the value for the column index 5 is equal to 3 (LaGuardia Airport), assign the value 1 to column index 15.
* For rows where the value for the column index 5 is equal to 5 (Newark Airport), assign the value 1 to column index 15.

In [44]:
taxi_modified=taxi.copy()

# create a new column filled with `0`.
import numpy as np 
zeros = np.zeros([taxi.shape[0], 1])
taxi_modified = np.concatenate([taxi, zeros], axis=1)

In [45]:
print(taxi_modified)

[[2.016e+03 1.000e+00 1.000e+00 ... 6.999e+01 1.000e+00 0.000e+00]
 [2.016e+03 1.000e+00 1.000e+00 ... 5.430e+01 1.000e+00 0.000e+00]
 [2.016e+03 1.000e+00 1.000e+00 ... 3.780e+01 2.000e+00 0.000e+00]
 ...
 [2.016e+03 6.000e+00 3.000e+01 ... 6.334e+01 1.000e+00 0.000e+00]
 [2.016e+03 6.000e+00 3.000e+01 ... 4.475e+01 1.000e+00 0.000e+00]
 [2.016e+03 6.000e+00 3.000e+01 ... 5.484e+01 2.000e+00 0.000e+00]]


In [46]:
taxi_modified[taxi_modified[:,5]==2,15]=1
taxi_modified[taxi_modified[:,5]==3,15]=1
taxi_modified[taxi_modified[:,5]==5,15]=1

# 9. Challenge: Which is the most popular airport?

### To complete this task, we'll need to check if the dropoff_location_code column (column index 6) is equal to one of the following values:

* 2: JFK Airport
* 3: LaGuardia Airport
* 5: Newark Airport.

In [47]:
jfk=taxi[taxi[:,6]==2]
LaG=taxi[taxi[:,6]==3]
Newark=taxi[taxi[:,6]==5]

In [48]:
jfk.shape[0]

11832

In [50]:
LaG.shape[0]

16602

In [52]:
Newark.shape[0]

63

### Observation:
LaGuardia Airport is most popular airport

# 10. Challenge: Calculating Statistics for Trips on Clean Data

#### The columns we're interested in are:

* trip_distance, at column index 7
* trip_length, at column index 8
* total_amount, at column index 13

In [53]:
# Calculate trip per miles
trip_mph = taxi[:,7] / (taxi[:,8] / 3600)

### TO DO:
* Create a new ndarray, cleaned_taxi, containing only rows for which the values of trip_mph are less than 100.
* Calculate the mean of the trip_distance column of cleaned_taxi. Assign the result to mean_distance.
* Calculate the mean of the trip_length column of cleaned_taxi. Assign the result to mean_length.
* Calculate the mean of the total_amount column of cleaned_taxi. Assign the result to mean_total_amount.

In [57]:
cleaned_taxi=taxi[trip_mph<100]

mean_distance=cleaned_taxi[:,7].mean()
mean_distance

12.666396599932893

In [58]:
mean_length=cleaned_taxi[:,8].mean()
mean_length

2239.503657309026

In [61]:
mean_total_amount=cleaned_taxi[:,13].mean()
mean_total_amount

48.98131853260262

## In this mission we learned:

* How to use numpy.genfromtxt() to read in an ndarray.
* About NaN values.
* What a boolean array is, and how to create one.
* How to use boolean indexing to filter values in one and two-dimensional ndarrays.
* How to assign one or more new values to an ndarray based on their locations.
* How to assign one or more new values to an ndarray based on their values.